specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,440 @@
+power adjustments for the new delegatee set using `_newDelegations`. Afterwards, the function MUST update
+the voting power of the current and new set of delegatees. Throughout this process, the function MUST use the
+`DelegationAdjustment` and `Op` types to store voting power adjustments and the operation to perform (add or
+subtract voting power).
+
+The function MUST then sort the new delegatee set in descending order, and store the delegations as part of the
+`_delegations` mapping. While sorting, the function MUST check for the validity of the delegations, as specified
+in the [Delegation Validation](#delegation-validation) section. Additionally, the function MUST emit a
+`DelegateVotesChanged` event for each change in delegatee voting power.
+
+### Getters
+
+For backwards compatibility, the `GovernanceDelegation` contract MUST implement all public getter functions of the
+`GovernanceToken` related to delegation and voting power. These functions MUST be used by the
+`GovernanceToken` when an account has been migrated to the `GovernanceDelegation` contract. Otherwise,
+the `GovernanceToken` MUST use its own state. Similarly, all of the `GovernanceDelegation` getter functions
+MUST use the `GovernanceToken` state if the account has not been migrated.
+
+#### `checkpoints`
+
+Retrieves the checkpoints for a given user address.
+
+```solidity
+function checkpoints(address _account) external view returns (ERC20Votes.Checkpoint[] memory)
+```
+
+#### `numCheckpoints`
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `delegates`
+
+Retrieves the delegation of a given user address with the highest voting power.
+This function is intended to be used by the `GovernanceToken` contract to maximize for backwards compatibility.
+
+```solidity
+function delegates(address _account) external view returns (address)
+```
+
+#### `getVotes`
+
+Retrieves the voting power of a given user address.
+
+```solidity
+function getVotes(address _account) external view returns (uint256)
+```
+
+#### `getPastVotes`
+
+Retrieves the voting power of a given user address at a block.
+
+```solidity
+function getPastVotes(address _account, uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `getPastTotalSupply`
+
+Retrieves the total supply of the `GovernanceToken` at a given block.
+
+```solidity
+function getPastTotalSupply(uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `migrated`
+
+Returns the migration status of an account — `True` if the account has been migrated, `False` otherwise.
+
+```solidity
+function migrated(address _account) public view returns (bool)
+```
+
+#### `delegations`
+
+Retrieves the delegations of a given user address sorted in descending order by voting power.
+
+```solidity
+function delegations(address _account) public view returns (Delegation[] memory)
+```
+
+### Events
```

Comment:
If delegations are keyed by L1 block number then we should include the L1 block number as part of these events. L2 block number can already be retrieved from events by default but I suspect having the L1 block number in there will be useful too.

Reply:
@smartcontracts i think would be part of the interop specs

Reply:
I agree, closing this comment for now and will discuss in the OP interop design doc.